source/byte_buf.c

```diff
@@ -162,48 +163,44 @@ bool aws_byte_cursor_next_split(
     char split_on,
     struct aws_byte_cursor *AWS_RESTRICT substr) {
 
-    bool first_run = false;
-    if (!substr->ptr) {
-        first_run = true;
-        substr->ptr = input_str->ptr;
-        substr->len = 0;
-    }
-
-    if (substr->ptr > input_str->ptr + input_str->len) {
-        /* This will hit if the last substring returned was an empty string after terminating split_on. */
-        AWS_ZERO_STRUCT(*substr);
-        return false;
-    }
-
-    /* Calculate first byte to search. */
-    substr->ptr += substr->len;
-    /* Remaining bytes is the number we started with minus the number of bytes already read. */
-    substr->len = input_str->len - (substr->ptr - input_str->ptr);
-
-    if (!first_run && substr->len == 0) {
-        /* This will hit if the string doesn't end with split_on but we're done. */
-        AWS_ZERO_STRUCT(*substr);
-        return false;
-    }
-
-    if (!first_run && *substr->ptr == split_on) {
-        /* If not first rodeo and the character after substr is split_on, skip. */
-        ++substr->ptr;
-        --substr->len;
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(input_str));
+
+    /* If substr is zeroed-out, then this is the first run. */
+    const bool first_run = substr->ptr == NULL;
+    if (first_run) {
+        if (input_str->ptr) {
+            *substr = *input_str;
+        } else {
+            /* NOTE: It's also legal for input_str to be zeroed out.
+             * In that case, set substr->ptr to something else so the next split() call doesn't look like first run. */
+            substr->ptr = (void *)"";
```

Comment:
Extreme case: what if the input_str also points to `(void*)""`

Reply:
That would be fine. In that case we'd take the `if (input_str->ptr)` branch above.
we just need `substr->ptr` to point to something non-null